cpp/src/arrow/util/iterator_test.cc

```diff
@@ -940,4 +942,144 @@ TEST(TestAsyncIteratorTransform, SkipSome) {
   AssertAsyncGeneratorMatch({"1", "3"}, std::move(filtered));
 }
 
+TEST(PushGenerator, Empty) {
+  PushGenerator<TestInt> gen;
+  auto producer = gen.producer();
+
+  auto fut = gen();
+  AssertNotFinished(fut);
+  producer.Close();
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), fut);
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), gen());
+
+  // Close idempotent
+  fut = gen();
+  producer.Close();
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), fut);
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), gen());
+  ASSERT_FINISHES_OK_AND_EQ(IterationTraits<TestInt>::End(), gen());
```

Comment:
I feel like this check might be unnecessary?  Can't hurt though.

Reply:
Perhaps over-cautious :-)